lib/render/achievement.php

```diff
@@ -1,127 +1,108 @@
-        $smallBadge = "<img loading='lazy' width='$smallBadgeSize' height='$smallBadgeSize' src='" . media_asset($smallBadgePath) . "' alt='$achNameAttr' title='$achNameAttr' class='$imgclass'>";
+        if ($label !== false) {
+            $title = $achievement['AchievementTitle'] ?? $achievement['Title'];
+            $points = $achievement['Points'] ?? null;
+            $label = $title . ($points ? ' (' . $points . ')' : '');
+        }
+
+        if ($icon !== false) {
+            $badgeName = is_string($icon) ? $icon : $achievement['BadgeName'] ?? null;
+            $icon = media_asset("/Badge/$badgeName.png");
+        }
 
-        if ($smallBadgeOnly) {
-            $displayable = "";
-        } else {
-            $smallBadge .= ' ';
+        if ($achievement['HardcoreMode'] ?? false) {
+            $iconClass = 'goldimage';
         }
+
+        // pre-render tooltip
+        $tooltip = $tooltip !== false ? $achievement : false;
     }
 
-    return "<div class='inline' onmouseover=\"Tip(loadCard('achievement', $achID))\" onmouseout=\"UnTip()\" >" .
-        "<a href='/achievement/$achID'>" .
-        "$smallBadge" .
-        "$displayable" .
-        "</a>" .
-        "</div>";
+    return avatar(
+        resource: 'achievement',
+        id: $id,
+        label: $label !== false && ($label || !$icon) ? $label : null,
+        link: route('achievement.show', $id),
+        tooltip: is_array($tooltip) ? renderAchievementCard($tooltip) : $tooltip,
+        iconUrl: $icon !== false && ($icon || !$label) ? $icon : null,
+        iconSize: $iconSize,
+        iconClass: $iconClass,
+        context: $context,
+    );
 }
 
-function renderAchievementCard(int $achievementId): string
+function renderAchievementCard(int|string|array $achievement, ?string $context = null): string
 {
-    $achData = [];
-    getAchievementMetadata($achievementId, $achData);
+    $id = is_int($achievement) || is_string($achievement) ? (int) $achievement : ($achievement['AchievementID'] ?? $achievement['ID'] ?? null);
+
+    if (empty($id)) {
+        return __('legacy.error.error');
+    }
+
+    $data = [];
+    if (is_array($achievement)) {
+        $data = $achievement;
+    }
+
+    if (empty($data)) {
+        $data = Cache::store('array')->rememberForever('achievement:' . $id . ':card-data', function () use ($id) {
+            $data = [];
+            getAchievementMetadata($id, $data);
+
+            return $data;
+        });
+    }
 
-    $badgeName = $achData['BadgeName'];
-    $achNameStr = $achData['AchievementTitle'];
-    $achDescStr = $achData['Description'];
-    $gameNameStr = $achData['GameTitle'];
-    $achPoints = $achData['Points'];
+    // /** @var ?User $user */
+    // $user = request()->user();
+    // if ($context && $user) {
+    //     getAchievementUnlocksData($id, $numWinners, $numPossibleWinners, $numRecentWinners, winnerInfo: $data, user: $user->User);
+    //
+    //     $data['Unlock'] = 'UNLOCKED';
+    // }
```

Comment:
commented code can be removed. looks to be handled below

Reply:
It was meant to allow for fetching unlock data in cases where no data is given for rendering the tooltip upfront. The `context` parameter would take care of that, however I did not find time to finish it, and it wasn't needed either as we pass said data along in all areas where it's needed. I'd like to keep it for now and see if i can finish it; passing the user context might come in handy in the future, not only for fetching unlock data for achievements asynchronously but for other resources types as well where user context can be applied.

I had a similar thing prepared in v2 where the user context for e.g. games would show additional user-specific infos.

Reply:
Removed - I'll revisit that feature another time.